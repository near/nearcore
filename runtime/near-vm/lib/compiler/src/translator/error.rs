use crate::{CompileError, WasmError};
use wasmparser::BinaryReaderError;

/// Return an `Err(WasmError::Unsupported(msg))` where `msg` the string built by calling `format!`
/// on the arguments to this macro.
#[macro_export]
macro_rules! wasm_unsupported {
    ($($arg:tt)*) => { $crate::WasmError::Unsupported(format!($($arg)*)) }
}

impl From<BinaryReaderError> for WasmError {
    fn from(original: BinaryReaderError) -> Self {
        Self::InvalidWebAssembly { message: original.message().into(), offset: original.offset() }
    }
}

impl From<BinaryReaderError> for CompileError {
    fn from(original: BinaryReaderError) -> Self {
        // `From` does not seem to be transitive by default, so we convert
        // BinaryReaderError -> WasmError -> CompileError
        Self::from(WasmError::from(original))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use wasmparser::BinaryReader;

    #[test]
    fn can_convert_binary_reader_error_to_wasm_error() {
        let mut reader = BinaryReader::new(b"\0\0\0\0");
        let binary_reader_error = reader.read_bytes(10).unwrap_err();
        match WasmError::from(binary_reader_error) {
            WasmError::InvalidWebAssembly { message, offset } => {
                assert_eq!(message, "unexpected end-of-file");
                assert_eq!(offset, 0);
            }
            err => panic!("Unexpected error: {:?}", err),
        }
    }

    #[test]
    fn can_convert_binary_reader_error_to_compile_error() {
        let mut reader = BinaryReader::new(b"\0\0\0\0");
        let binary_reader_error = reader.read_bytes(10).unwrap_err();
        match CompileError::from(binary_reader_error) {
            CompileError::Wasm(WasmError::InvalidWebAssembly { message, offset }) => {
                assert_eq!(message, "unexpected end-of-file");
                assert_eq!(offset, 0);
            }
            err => panic!("Unexpected error: {:?}", err),
        }
    }
}
