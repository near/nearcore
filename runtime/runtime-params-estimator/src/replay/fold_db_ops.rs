use super::cache_stats::CacheStats;
use super::Visitor;
use std::collections::{BTreeMap, HashMap};
use std::io::Write;

const EMPTY_STATE_ERR: &str = "states must never be empty";

/// A visitor that keeps track of DB operations and aggregates it by specific labels.
pub(super) struct FoldDbOps {
    /// Labels at which to break aggregation, and the fields to print on each.
    ///
    /// If field is not available, it will be silently ignored.
    fold_anchors: HashMap<String, Vec<String>>,
    /// Print the DB operations that are at indentation 0, not associated with a span.
    ///
    /// This is a summary of the total. Instead it shows what operations have
    /// not been covered by already reported output. To get a summary of the
    /// total, run without folding.
    print_top_level: bool,
    /// Only show data for a specific smart contract, specified by account id.
    /// Currently only applies within receipts.
    account_filter: Option<String>,
    /// Only evaluate lines with at least this depth of indentation.
    ///
    /// Used to apply filters and skip some parts of the trace.
    /// Default is 0, which collects everything.
    min_indent: usize,
    /// Reset the filter when indentation goes back to this value.
    ///
    /// Used in combination with `min_indent` and `account_filter`.
    /// Resetting means that `min_indent` is set to usize::MAX and nothing will
    /// be evaluated until the `account_filter` is triggered again and sets it
    /// to a smaller value again.
    filter_reset_indent: Option<usize>,
    /// Optionally collect and print detailed statistics for cache hits and misses.
    track_caches: bool,
    /// Keeps track of current block.
    block_hash: Option<String>,
    /// Stack of states, each starting at a specific indent.
    states: Vec<State>,
}

/// Stores statistics for a range of the trace, starting at a fixed indent and
/// collecting data for everything indented further, until another state object
/// is pushed on top.
#[derive(Default)]
struct State {
    /// The indent at which this state started
    indent: usize,
    /// Keeps track of operations per DB column.
    ops_cols: BTreeMap<String, BTreeMap<String, usize>>,
    /// Optionally collect and print detailed statistics for cache hits and misses.
    cache_stats: Option<CacheStats>,
}

impl FoldDbOps {
    pub(super) fn new() -> Self {
        Self {
            fold_anchors: HashMap::new(),
            print_top_level: true,
            account_filter: None,
            track_caches: false,
            states: vec![State::default()],
            block_hash: None,
            min_indent: 0,
            filter_reset_indent: None,
        }
    }

    /// Pre-set that folds on chunks.
    pub(super) fn chunks(self) -> Self {
        self.fold("apply_transactions", &["receiver", "shard_id"])
    }

    /// Pre-set that folds on receipts.
    pub(super) fn receipts(self) -> Self {
        self.fold("process_receipt", &["receiver", "receipt_id"])
            .fold("process_transaction", &["tx_hash"])
            .print_top_level(false)
    }

    pub(super) fn fold(mut self, anchor: impl Into<String>, printed_fields: &[&str]) -> Self {
        let fields = printed_fields.into_iter().map(|s| (*s).into()).collect();
        self.fold_anchors.insert(anchor.into(), fields);
        self
    }

    pub(super) fn print_top_level(mut self, yes: bool) -> Self {
        self.print_top_level = yes;
        self
    }

    pub(super) fn with_cache_stats(mut self) -> Self {
        self.track_caches = true;
        self.state().cache_stats = Some(CacheStats::default());
        self
    }

    pub(super) fn account_filter(mut self, account: Option<String>) -> Self {
        if account.is_some() {
            // evaluate nothing if there is a filter, until the filter matches the first time
            self.min_indent = usize::MAX;
        }
        self.account_filter = account;
        self
    }

    fn state(&mut self) -> &mut State {
        self.states.last_mut().expect(EMPTY_STATE_ERR)
    }

    fn push_state(&mut self, indent: usize) {
        let cache_stats = if self.track_caches { Some(CacheStats::default()) } else { None };
        let new_state = State { indent, ops_cols: Default::default(), cache_stats };
        self.states.push(new_state);
    }

    fn pop_state(&mut self) -> State {
        let state = self.states.pop().expect(EMPTY_STATE_ERR);
        if self.states.is_empty() {
            self.push_state(0);
        }
        state
    }

    fn skip_eval(&mut self, trace_indent: usize) -> bool {
        trace_indent < self.min_indent
    }

    /// Check if indentation has gone back enough to pop current state or reset filter.
    ///
    /// Call this before `skip()` to ensure it uses the correct `min_indent`.
    fn update_state(&mut self, out: &mut dyn Write, indent: usize) -> anyhow::Result<()> {
        if self.states.len() > 1 && self.state().indent >= indent {
            self.pop_state().print(out)?;
        }
        if let Some(reset_indent) = self.filter_reset_indent {
            if indent <= reset_indent {
                self.filter_reset_indent = None;
                self.min_indent = usize::MAX;
            }
        }
        Ok(())
    }
}

impl Visitor for FoldDbOps {
    fn eval_db_op(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        op: &str,
        size: Option<u64>,
        key: &[u8],
        col: &str,
    ) -> anyhow::Result<()> {
        self.update_state(out, indent)?;

        // Block hash in traces is visibly by looking at DB lookups of
        // `BlockInfo` column. Keep track of it so that each time something is
        // printed, the block hash can be included if desired.
        match op {
            "GET" => {
                if col == "BlockInfo" {
                    self.block_hash = Some(bs58::encode(key).into_string());
                }
            }
            _ => {
                // nop
            }
        }

        if self.skip_eval(indent) {
            return Ok(());
        }

        // Count DB operation for the corresponding column.
        *self
            .state()
            .ops_cols
            .entry(op.to_owned())
            .or_default()
            .entry(col.to_owned())
            .or_default() += 1;

        if let Some(cache_stats) = &mut self.state().cache_stats {
            cache_stats.eval_db_op(op, size);
        }

        Ok(())
    }

    fn eval_storage_op(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        op: &str,
        dict: &BTreeMap<&str, &str>,
    ) -> anyhow::Result<()> {
        self.update_state(out, indent)?;
        if self.skip_eval(indent) {
            return Ok(());
        }
        if let Some(cache_stats) = &mut self.state().cache_stats {
            cache_stats.eval_storage_op(op, dict)?;
        }
        Ok(())
    }

    fn eval_label(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        label: &str,
        dict: &BTreeMap<&str, &str>,
    ) -> anyhow::Result<()> {
        self.update_state(out, indent)?;
        if let (Some(receiver), Some(filtered_account)) =
            (dict.get("receiver"), &self.account_filter)
        {
            if receiver == filtered_account && self.filter_reset_indent.is_none() {
                self.min_indent = indent;
                self.filter_reset_indent = Some(indent);
            }
        }
        if self.skip_eval(indent) {
            return Ok(());
        }
        if self.fold_anchors.contains_key(label) {
            // Section to fold on starts. Push a new state on the stack and
            // print the header for the new section.
            self.push_state(indent);
            write!(out, "{:indent$}{label}", "")?;
            // Unnecessary perf optimization: Second lookup in fold anchors
            // could be avoided by reading the key directly but then we keep
            // a mutable reference to self and cannot naively call
            // self.push_state above.
            // Better to lookup twice and keep code simple.
            for key in self.fold_anchors.get(label).expect("just checked contains key").iter() {
                if let Some(value) = dict.get(key.as_str()) {
                    write!(out, " {key}={value}")?;
                }
            }
            if let Some(block) = &self.block_hash {
                write!(out, " block={block}")?;
            }
            writeln!(out)?;
        }

        if let Some(cache_stats) = &mut self.state().cache_stats {
            cache_stats.eval_generic_label(dict);
        }
        Ok(())
    }

    fn flush(&mut self, out: &mut dyn Write) -> anyhow::Result<()> {
        if self.print_top_level {
            writeln!(out, "top-level:")?;
            self.pop_state().print(out)?;
        }
        Ok(())
    }
}

impl State {
    fn print(self, out: &mut dyn Write) -> anyhow::Result<()> {
        let indent = self.indent + 2;
        for (op, map) in self.ops_cols.into_iter() {
            if !map.is_empty() {
                write!(out, "{:indent$}{op}   ", "")?;
            }
            for (col, num) in map.into_iter() {
                write!(out, "{num:8>} {col}  ")?;
            }
            writeln!(out)?;
        }

        if let Some(stats) = self.cache_stats {
            stats.print(out, indent)?;
        }
        writeln!(out)?;
        Ok(())
    }
}
