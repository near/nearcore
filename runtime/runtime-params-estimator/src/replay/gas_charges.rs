use super::Visitor;
use std::collections::BTreeMap;
use std::io::Write;

/// Visitor that tracks which DB operations are charged with gas and which are
/// for free.
#[derive(Default)]
pub(super) struct ChargedVsFree {
    in_fn_call: bool,
    storage_op_indent: usize,

    free_gets: u64,
    free_gets_size: u64,
    charged_gets: u64,
    charged_gets_size: u64,
}

impl Visitor for ChargedVsFree {
    fn eval_db_op(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        op: &str,
        size: Option<u64>,
        _key: &[u8],
        _col: &str,
    ) -> anyhow::Result<()> {
        self.eval_label(out, indent, op, &BTreeMap::new())?;
        if op != "GET" {
            return Ok(());
        }
        if self.in_fn_call {
            self.charged_gets += 1;
            self.charged_gets_size += size.unwrap_or(0);
        } else {
            self.free_gets += 1;
            self.free_gets_size += size.unwrap_or(0);
        }
        Ok(())
    }

    fn eval_storage_op(
        &mut self,
        _out: &mut dyn Write,
        indent: usize,
        _op: &str,
        _dict: &BTreeMap<&str, &str>,
    ) -> anyhow::Result<()> {
        self.storage_op_indent = indent + 2;
        self.in_fn_call = true;
        Ok(())
    }

    fn eval_label(
        &mut self,
        _out: &mut dyn Write,
        indent: usize,
        _label: &str,
        _dict: &BTreeMap<&str, &str>,
    ) -> anyhow::Result<()> {
        if indent < self.storage_op_indent {
            self.in_fn_call = false;
        }
        Ok(())
    }

    fn flush(&mut self, out: &mut dyn Write) -> anyhow::Result<()> {
        writeln!(
            out,
            "{:>8} free gets with total size of    {:>8}",
            self.free_gets, self.free_gets_size
        )?;
        writeln!(
            out,
            "{:>8} charged gets with total size of {:>8}",
            self.charged_gets, self.charged_gets_size
        )?;
        writeln!(
            out,
            "{:>7.2}% of gets uncharged",
            100.0 * self.free_gets as f64 / (self.charged_gets + self.free_gets) as f64
        )?;
        writeln!(
            out,
            "{:>7.2}% of gets total size uncharged",
            100.0 * self.free_gets_size as f64
                / (self.charged_gets_size + self.free_gets_size) as f64
        )?;

        self.free_gets = 0;
        self.free_gets_size = 0;
        self.charged_gets = 0;
        self.charged_gets_size = 0;

        Ok(())
    }
}
