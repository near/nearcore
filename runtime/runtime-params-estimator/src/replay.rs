use anyhow::Context;
use std::collections::BTreeMap;
use std::fs::File;
use std::io::{self, Write};
use std::path::PathBuf;
use std::str::SplitWhitespace;
use tracing::log::error;

use self::fold_db_ops::FoldDbOps;
use self::gas_charges::ChargedVsFree;

mod cache_stats;
mod fold_db_ops;
mod gas_charges;

#[derive(clap::Parser)]
pub(crate) struct ReplayCmd {
    trace: PathBuf,
    #[clap(subcommand)]
    mode: ReplayMode,
    /// Only show data for a specific smart contract, specified by account id.
    #[clap(long)]
    account: Option<String>,
}

#[derive(Clone, Copy, clap::Subcommand, Debug)]
pub(crate) enum ReplayMode {
    /// Print DB accesses and cache statistics for the entire trace.
    CacheStats,
    /// Print DB accesses per receipt.
    ReceiptDbStats,
    /// Print DB accesses and cache statistics per receipt.
    ReceiptCacheStats,
    /// Print DB accesses per chunk.
    ChunkDbStats,
    /// Print DB accesses and cache statistics per chunk.
    ChunkCacheStats,
    /// Go over DB operations and print how much of it is paid for with gas.
    GasCharges,
}

impl ReplayCmd {
    pub(crate) fn run(&self, out: &mut dyn Write) -> anyhow::Result<()> {
        let file = File::open(&self.trace)?;
        self.run_on_input(io::BufReader::new(file), out)
    }

    fn run_on_input(&self, input: impl io::BufRead, out: &mut dyn Write) -> anyhow::Result<()> {
        let mut visitor = self.build_visitor();
        for line in input.lines() {
            let line = line?;
            if let Err(e) = visitor.eval_line(out, &line) {
                error!("ERROR: {e} for input line: {line}");
            }
        }
        visitor.flush(out)?;
        Ok(())
    }

    fn build_visitor(&self) -> Box<dyn Visitor> {
        match &self.mode {
            ReplayMode::CacheStats => {
                Box::new(FoldDbOps::new().with_cache_stats().account_filter(self.account.clone()))
            }
            ReplayMode::ChunkDbStats => {
                if self.account.is_some() {
                    unimplemented!("account filter does not work with per-chunk statistics");
                }
                Box::new(FoldDbOps::new().chunks())
            }
            ReplayMode::ChunkCacheStats => {
                if self.account.is_some() {
                    unimplemented!("account filter does not work with per-chunk statistics");
                }
                Box::new(FoldDbOps::new().chunks().with_cache_stats())
            }
            ReplayMode::ReceiptDbStats => {
                Box::new(FoldDbOps::new().receipts().account_filter(self.account.clone()))
            }
            ReplayMode::ReceiptCacheStats => Box::new(
                FoldDbOps::new().receipts().with_cache_stats().account_filter(self.account.clone()),
            ),
            ReplayMode::GasCharges => {
                if self.account.is_some() {
                    unimplemented!("account filter does not work with gas charges");
                }
                Box::new(ChargedVsFree::default())
            }
        }
    }
}

/// Interface for processing an IO trace line-by-line.
///
/// The visitor default methods take over the basic parsing of the IO trace.
/// Specific implementations can extract the data of interest and process it
/// further as necessary, without duplicating the parsing code.
trait Visitor {
    /// The root entry point of the visitors.
    ///
    /// This function takes a raw input line as input without any preprocessing.
    /// A visitor may choose to overwrite this function for full control but the
    /// intention is that the default implementation takes over the basic
    /// parsing and visitor implementations define their behaviour using the
    /// other trait methods.
    fn eval_line(&mut self, out: &mut dyn Write, line: &str) -> anyhow::Result<()> {
        if let Some(indent) = line.chars().position(|c| !c.is_whitespace()) {
            let mut tokens = line.split_whitespace();
            if let Some(keyword) = tokens.next() {
                match keyword {
                    "GET" | "SET" | "UPDATE_RC" => {
                        let col = tokens.next().context("missing column field in DB operation")?;
                        let mut key_str = tokens.next().context("missing key in DB operation")?;
                        if key_str.starts_with('"') {
                            key_str = &key_str[1..key_str.len() - 1];
                        }
                        let key = near_fmt::Bytes::from_str(key_str).map_err(|e| {
                            anyhow::anyhow!("failed to decode key {key_str} because {e}")
                        })?;
                        let dict = extract_key_values(tokens)?;
                        let size: Option<u64> = dict.get("size").map(|s| s.parse()).transpose()?;
                        self.eval_db_op(out, indent, keyword, size, &key, col)?;
                    }
                    "storage_read" | "storage_write" | "storage_remove" | "storage_has_key" => {
                        let op = tokens.next();
                        if op.is_none() {
                            return Ok(());
                        }

                        let dict = extract_key_values(tokens)?;
                        self.eval_storage_op(out, indent, keyword, &dict)?;
                    }
                    other_label => {
                        let dict = extract_key_values(tokens)?;
                        self.eval_label(out, indent, other_label, &dict)?;
                    }
                }
            }
        }
        Ok(())
    }

    /// Gets called for every trie storage operation.
    ///
    /// A storage operation is a layer above DB operations, one storage
    /// operation can cause many DB operations, or it could be resolved from
    /// cache without any DB operations.
    fn eval_storage_op(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        op: &str,
        dict: &BTreeMap<&str, &str>,
    ) -> anyhow::Result<()> {
        let (_, _, _, _) = (out, indent, op, dict);
        Ok(())
    }

    /// Gets called for every DB operation.
    fn eval_db_op(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        op: &str,
        size: Option<u64>,
        key: &[u8],
        col: &str,
    ) -> anyhow::Result<()> {
        if col == "State" {
            self.eval_state_db_op(out, indent, op, size, key)
        } else {
            Ok(())
        }
    }

    /// Gets called for every DB operation on the state column.
    fn eval_state_db_op(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        op: &str,
        size: Option<u64>,
        key: &[u8],
    ) -> anyhow::Result<()> {
        let (_, _, _, _, _) = (out, indent, op, size, key);
        Ok(())
    }

    /// Opening spans that are not storage or DB operations.
    fn eval_label(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        label: &str,
        dict: &BTreeMap<&str, &str>,
    ) -> anyhow::Result<()> {
        let (_, _, _, _) = (out, indent, label, dict);
        Ok(())
    }

    /// Print accumulated internal state.
    ///
    /// This function is called at the end of a replayed trace. Visitor
    /// implementation usually choose to call this more often. For example, once
    /// for each receipt.
    fn flush(&mut self, out: &mut dyn Write) -> anyhow::Result<()> {
        let _ = out;
        Ok(())
    }
}

fn extract_key_values<'a>(
    mut tokens: SplitWhitespace<'a>,
) -> anyhow::Result<BTreeMap<&'a str, &'a str>> {
    let mut dict = BTreeMap::new();
    while let Some(key_val) = tokens.next() {
        let (key, value) =
            key_val.split_once('=').context("key-value pair delimited by `=` expected")?;
        dict.insert(key, value);
    }
    Ok(dict)
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use super::{ReplayCmd, ReplayMode};

    /// These inputs are real mainnet traffic for the given block heights.
    /// Each trace contains two chunks in one shard.
    /// In combination, they cover an interesting variety of content.
    ///  Shard 0: Many receipts, but no function calls.
    ///  Shard 1: Empty chunks without txs / receipts.
    ///  Shard 2: Some receipts and txs, but no function calls.
    ///  Shard 3: Many function calls with lots of storage accesses.
    const INPUT_TRACES: &[&str] = &[
        "75220100-75220101.s0.io_trace",
        "75220100-75220101.s1.io_trace",
        "75220100-75220101.s2.io_trace",
        "75220100-75220101.s3.io_trace",
    ];

    /// A synthetic trace that should be a bit easier to read and verify.
    const SYNTHETIC_TRACE: &str = r#"
GET BlockHeader "`fAkeHeAd3R`" size=6000
GET BlockInfo "`FAk31nf0`" size=1
apply_transactions shard_id=0
    process_state_update 
        apply num_transactions=1 shard_cache_hit=10 shard_cache_miss=1
            process_transaction tx_hash=txHash0 shard_cache_miss=1 shard_cache_hit=20
                GET State "'stateKey0'" size=300
            process_receipt receipt_id=id0 predecessor=system receiver=alice.near id=id0 shard_cache_miss=2 shard_cache_hit=6
                GET State "'stateKey1'" size=10
                GET State "'stateKey2'" size=20
            process_receipt receipt_id=id1 predecessor=system receiver=bob.near id=id1 shard_cache_hit=18 shard_cache_miss=3 shard_cache_too_large=1
                GET State "'stateKey3'" size=100
                GET State "'stateKey4'" size=200
                GET State "'stateKey5'" size=9000
            process_receipt receipt_id=id2 predecessor=system receiver=alice.near id=id2 shard_cache_miss=1 shard_cache_hit=6
                GET State "'stateKey6'" size=30
                attached_deposit 
                input 
                register_len 
                read_register 
                storage_read READ key=StorageKey0 size=1000 tn_db_reads=20 tn_mem_reads=0 shard_cache_hit=19 shard_cache_miss=1
                    GET State "'stateKey7'" size=5
            process_receipt receipt_id=id3 predecessor=utiha.near receiver=alice.near id=id3 shard_cache_miss=0 shard_cache_hit=15
            GET State "'stateKey8'" size=300
        GET State "'stateKey9'" size=400
GET State "'stateKey10'" size=500
"#;

    #[test]
    fn test_cache_stats() {
        check_replay_mode(ReplayMode::CacheStats);
    }

    #[test]
    fn test_chunk_cache_stats() {
        check_replay_mode(ReplayMode::ChunkCacheStats);
    }

    #[test]
    fn test_chunk_db_stats() {
        check_replay_mode(ReplayMode::ChunkDbStats);
    }

    #[test]
    fn test_gas_charges() {
        check_replay_mode(ReplayMode::GasCharges);
    }

    #[test]
    fn test_receipt_cache_stats() {
        check_replay_mode(ReplayMode::ReceiptCacheStats);
    }

    #[test]
    fn test_receipt_db_stats() {
        check_replay_mode(ReplayMode::ReceiptDbStats);
    }

    #[track_caller]
    fn check_replay_mode(mode: ReplayMode) {
        for trace_name in INPUT_TRACES {
            let dir = env!("CARGO_MANIFEST_DIR");
            let trace_path = std::path::Path::new(dir).join("res").join(trace_name);
            let cmd = ReplayCmd { trace: trace_path, mode, account: None };
            let mut buffer = Vec::new();
            cmd.run(&mut buffer).unwrap_or_else(|e| {
                panic!("command should not fail for input {trace_name}, failure was {e}")
            });
            let output = String::from_utf8(buffer).unwrap_or_else(|e| {
                panic!("invalid output for input {trace_name}, failure was {e}")
            });
            insta::assert_snapshot!(format!("{mode:?}-{trace_name}"), output);
        }
    }

    #[test]
    fn test_account_filter_cache_stats() {
        check_account_filter(ReplayMode::CacheStats);
    }

    #[test]
    fn test_account_filter_receipt_cache_stats() {
        check_account_filter(ReplayMode::ReceiptCacheStats);
    }

    #[test]
    fn test_account_filter_receipt_db_stats() {
        check_account_filter(ReplayMode::ReceiptDbStats);
    }

    #[track_caller]
    fn check_account_filter(mode: ReplayMode) {
        let account = Some("alice.near".to_owned());
        // trace path not used, will be read from in-memory input instead
        let trace = PathBuf::new();
        let cmd = ReplayCmd { trace, mode, account };
        let mut buffer = Vec::new();
        cmd.run_on_input(SYNTHETIC_TRACE.as_bytes(), &mut buffer).expect("failed replaying");
        let output =
            String::from_utf8(buffer).unwrap_or_else(|e| panic!("invalid output, failure was {e}"));
        insta::assert_snapshot!(format!("account_filter_{mode:?}"), output);
    }
}
