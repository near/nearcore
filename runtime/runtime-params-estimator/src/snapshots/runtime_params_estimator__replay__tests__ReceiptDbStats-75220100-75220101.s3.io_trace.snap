---
source: runtime/runtime-params-estimator/src/replay.rs
expression: output
---
      process_transaction tx_hash=AA7U5AQwKV5zCMEqN1317cqwBQkAMRiWtAtoRSNjXEjT block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
        GET   32 State  

      process_transaction tx_hash=AKtUBGwV1zxbG3dRjCQD9WJ12Fzo9YgQ1ow2nZCd2m6t block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
        GET   21 State  

      process_transaction tx_hash=uSKvjryKgJQrWKg5EFSvEpnDUPRtRSbPrX42i84ekV3 block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
        GET   37 State  

      process_receipt receiver=verrelkeung.near receipt_id=cesi14ugy8KbNKu5SqL5SYY8MUFRWC39qzKYP9xiJL7 block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR

      process_receipt receiver=token.sweat receipt_id=4tSv2K8ohafiXnCAKprrWinzRFuaMwQzM83dkVx99q7r block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
        GET   70 State  

      process_receipt receiver=tge-lockup.sweat receipt_id=7UeSKKYxqRLsSwnW5xJm116QTJEWJJa7YqR1JmNWZE3L block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR

      process_receipt receiver=npo-aurora.near receipt_id=7HpTajwsXdnmX2EEzWq9PXktDJLGJGwqWhL4gcQ3QVQ3 block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
        GET   16 State  

      process_receipt receiver=tge-lockup.sweat receipt_id=CHPvzniPWKMWCiZ3c6ydqN6QgsEBpS3pNgP7HH48iQLK block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
        GET   20 State  

      process_receipt receiver=sweat_welcome.near receipt_id=H8GTFPKBkuUjvs8NfzcPNZuapFiMbGUMvscNZhnjUsPv block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
        GET   36 State  

      process_receipt receiver=token.sweat receipt_id=4rYKmJdE5pAHCDb9qVYRjTruvE58vRPRH9A7sGroBa3b block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
        GET   19 State  

      process_receipt receiver=token.sweat receipt_id=ohgxoczB4YhnAsUDpFVDpnMGP5iXuZbftgNe86KZ5XN block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
        GET   13 State  

      process_receipt receiver=token.sweat receipt_id=9iNcdK3djEA1mqNP4gaQM8jcaEvBtqas31S79cAWH4TY block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
        GET   8 State  

      process_receipt receiver=token.sweat receipt_id=22c3R5Xe82GCre4Mxa33Se7M4Rhp77RMf4ZwRjS55Ai7 block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
        GET   8 State  

      process_receipt receiver=ya_gul.near receipt_id=B2XmBAZhY3gYwktoRiP96KBip7Jd5iQNHbRXKLW2348W block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
        GET   16 State  

      process_receipt receiver=nex1234.near receipt_id=DbRHWcLYjARUjjUZj3ERR5SGcBE7nLrXTe1dKmuLWivu block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
        GET   14 State  

      process_receipt receiver=xtoken.ref-finance.near receipt_id=6RiUAxqBmiBo4XnEURb87D5twzmrZjELre4cJkEzdFuu block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
        GET   34 State  

      process_receipt receiver=tge-lockup.sweat receipt_id=EYwsRh17pfzg1n6uWEduBbrvbWmf2hnt2g8HPeABZLHR block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
        GET   25 State  

      process_transaction tx_hash=Fdaq5LbQw49MfMp1BHMfXCkDVA3RW9ZnYqTbTSrdz2VD block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9
        GET   44 State  

      process_transaction tx_hash=D8U1btrcfRbbf8wV1uWLmHKY1hDHBYUySwwE7hmPWjAP block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9
        GET   33 State  

      process_transaction tx_hash=FjwDNPxNxgpipRX3s48kRNUGfaPwgxG64N7KRQ7YULx2 block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9
        GET   24 State  

      process_transaction tx_hash=HN32RHYR9J1up2MguxxCpcyhnFwbxttKNTscT4n748SR block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9
        GET   20 State  

      process_transaction tx_hash=A76PNL2oTtuUtpe51dpCrcCU2NShF2vkB6ZTLHoXEnKx block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9

      process_receipt receiver=token.sweat receipt_id=8s88Y7hyAXrmZngJGpRBsvsbqYzAQiJbqdwSMESeLmr5 block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9
        GET   52 State  

      process_receipt receiver=seensayer.near receipt_id=J6m2SVXWwC3oR2g2QDbbqZf7GCTYx8tQETZuFCi1QDBU block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9
        GET   24 State  

      process_receipt receiver=tge-lockup.sweat receipt_id=AZawzkDrLTMrbxZTBWwgkYTY5AEcEBJQYsVrRckDayf4 block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9
        GET   27 State  

      process_receipt receiver=token.sweat receipt_id=454QLeWg3Yy4HQLcowLtvQGJfwzw2reUREx1sLtv1gga block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9
        GET   16 State  

      process_receipt receiver=tge-lockup.sweat receipt_id=Jf6Hh3cgANXPYkd536rGMFbWkXzkA1CBNFNJhwn3ksL block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9
        GET   6 State  

      process_receipt receiver=token.v2.ref-finance.near receipt_id=8vJXghN3eeQr2HLsSMEUmVb85sZymMSKseYNfKmK328o block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9
        GET   29 State  

      process_receipt receiver=xtoken.ref-finance.near receipt_id=9r96hCrbYcEzqaTRrA6mZHv26KJqrihjuomU6Hf6vfTX block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9

      process_receipt receiver=token.sweat receipt_id=7d1QXW1is9bzovZSA1XJK1AWTrPJPixNVReuCihPBQFh block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9
        GET   15 State  

      process_receipt receiver=tge-lockup.sweat receipt_id=EvN6kHZCYkBr4E8V8jm5R97Nf7paX6xhbtz57YZbHsn block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9


